File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_3_4_isp/before/TrainingSession.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_3_4_isp.before;

/**
 * ANTI-PATTERN: The "Fat Interface" (ISP Violation).
 * * ARCHITECTURE PROBLEM: The Interface Segregation Principle dictates that 
 * no client should be forced to depend on methods it does not use. 
 * * This interface is bloated. It assumes that anyone attending a training 
 * session needs to practice BOTH field skills (shooting, tackling) AND 
 * goalie skills (diving, hand distribution). This creates a massive 
 * burden on any specialized player trying to implement it.
 */
public interface TrainingSession {
    void practiceShooting();
    void practiceTackling();
    
    // 🚨 ARCHITECTURE WARNING: These methods pollute the interface 
    // for 90% of the players on the team!
    void practiceDivingSaves();
    void practiceHandDistribution();
}

File: Chapter 03/Python/section_3_3_4_isp/before/training_session.py
from abc import ABC, abstractmethod

"""
ANTI-PATTERN: The "Fat Interface" (ISP Violation).
* ARCHITECTURE PROBLEM: The TrainingSession ABC is trying to be everything 
to everyone. It bundles field player drills (shooting/tackling) with 
goalie-specific drills (diving/hands).
* This forces every player participating in a session to implement methods 
that are physically impossible or irrelevant for their role.
"""
class TrainingSession(ABC):
    @abstractmethod
    def practice_shooting(self): 
        """Valid for all players."""
        pass
    
    @abstractmethod
    def practice_tackling(self): 
        """Valid for field players."""
        pass
    
    @abstractmethod
    def practice_diving_saves(self): 
        """🚨 Goalie only!"""
        pass
    
    @abstractmethod
    def practice_hand_distribution(self): 
        """🚨 Goalie only!"""
        pass